File: src/state.rs
use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Comment {
    pub id: String,
    pub file_path: String,
    pub line_number: Option<usize>,
    pub text: String,
    pub timestamp: u64,
    pub branch: String,
    pub commit: String,
    pub resolved: bool,
}

#[derive(Serialize, Deserialize, Default)]
struct RepoState {
    viewed_files: Vec<String>,
    comments: Vec<Comment>,
}

#[derive(Serialize, Deserialize, Default)]
struct ViewedState {
    // repo_path -> branch -> commit -> RepoState
    repos: HashMap<String, HashMap<String, HashMap<String, RepoState>>>,
}

pub struct StateManager {
    state_file: PathBuf,
    state: ViewedState,
}

impl StateManager {
    pub fn new() -> Result<Self> {
        let state_dir = dirs::state_dir()
            .or_else(|| dirs::data_local_dir())
            .context("Failed to determine state directory")?
            .join("guck");

        fs::create_dir_all(&state_dir).context("Failed to create state directory")?;

        let state_file = state_dir.join("viewed.json");

        let state = if state_file.exists() {
            let contents = fs::read_to_string(&state_file).context("Failed to read state file")?;
            serde_json::from_str(&contents).unwrap_or_default()
        } else {
            ViewedState::default()
        };

        Ok(Self { state_file, state })
    }

    pub fn is_file_viewed(
        &self,
        repo_path: &str,
        branch: &str,
        commit: &str,
        file_path: &str,
    ) -> Result<bool> {
        Ok(self
            .state
            .repos
            .get(repo_path)
            .and_then(|branches| branches.get(branch))
            .and_then(|commits| commits.get(commit))
            .map(|repo_state| repo_state.viewed_files.contains(&file_path.to_string()))
            .unwrap_or(false))
    }

    pub fn mark_file_viewed(
        &mut self,
        repo_path: &str,
        branch: &str,
        commit: &str,
        file_path: &str,
    ) -> Result<()> {
        let repo = self
            .state
            .repos
            .entry(repo_path.to_string())
            .or_insert_with(HashMap::new);

        let branch_map = repo.entry(branch.to_string()).or_insert_with(HashMap::new);

        let repo_state = branch_map
            .entry(commit.to_string())
            .or_insert_with(RepoState::default);

        if !repo_state.viewed_files.contains(&file_path.to_string()) {
            repo_state.viewed_files.push(file_path.to_string());
        }

        self.save()?;
        Ok(())
    }

    pub fn unmark_file_viewed(
        &mut self,
        repo_path: &str,
        branch: &str,
        commit: &str,
        file_path: &str,
    ) -> Result<()> {
        if let Some(repo) = self.state.repos.get_mut(repo_path) {
            if let Some(branch_map) = repo.get_mut(branch) {
                if let Some(repo_state) = branch_map.get_mut(commit) {
                    repo_state.viewed_files.retain(|f| f != file_path);
                }
            }
        }

        self.save()?;
        Ok(())
    }

    pub fn add_comment(
        &mut self,
        repo_path: &str,
        branch: &str,
        commit: &str,
        file_path: &str,
        line_number: Option<usize>,
        text: String,
    ) -> Result<Comment> {
        use std::time::{SystemTime, UNIX_EPOCH};

        let repo = self
            .state
            .repos
            .entry(repo_path.to_string())
            .or_insert_with(HashMap::new);

        let branch_map = repo.entry(branch.to_string()).or_insert_with(HashMap::new);

        let repo_state = branch_map
            .entry(commit.to_string())
            .or_insert_with(RepoState::default);

        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs();

        let comment = Comment {
            id: format!("{}-{}", timestamp, repo_state.comments.len()),
            file_path: file_path.to_string(),
            line_number,
            text,
            timestamp,
            branch: branch.to_string(),
            commit: commit.to_string(),
            resolved: false,
        };

        repo_state.comments.push(comment.clone());
        self.save()?;
        Ok(comment)
    }

    pub fn get_comments(
        &self,
        repo_path: &str,
        branch: &str,
        commit: &str,
        file_path: Option<&str>,
    ) -> Result<Vec<Comment>> {
        let comments = self
            .state
            .repos
            .get(repo_path)
            .and_then(|branches| branches.get(branch))
            .and_then(|commits| commits.get(commit))
            .map(|repo_state| {
                if let Some(fp) = file_path {
                    repo_state
                        .comments
                        .iter()
                        .filter(|c| c.file_path == fp)
                        .cloned()
                        .collect()
                } else {
                    repo_state.comments.clone()
                }
            })
            .unwrap_or_default();

        Ok(comments)
    }

    pub fn resolve_comment(
        &mut self,
        repo_path: &str,
        branch: &str,
        commit: &str,
        comment_id: &str,
    ) -> Result<()> {
        if let Some(repo) = self.state.repos.get_mut(repo_path) {
            if let Some(branch_map) = repo.get_mut(branch) {
                if let Some(repo_state) = branch_map.get_mut(commit) {
                    if let Some(comment) =
                        repo_state.comments.iter_mut().find(|c| c.id == comment_id)
                    {
                        comment.resolved = true;
                    }
                }
            }
        }
        self.save()?;
        Ok(())
    }

    fn save(&self) -> Result<()> {
        let contents =
            serde_json::to_string_pretty(&self.state).context("Failed to serialize state")?;
        fs::write(&self.state_file, contents).context("Failed to write state file")?;
        Ok(())
    }
}

File: internal/state/state.go
package state

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"time"

	"github.com/tuist/guck/internal/config"
	"github.com/tuist/guck/internal/export"
)

type Comment struct {
	ID         string `json:"id"`
	FilePath   string `json:"file_path"`
	LineNumber *int   `json:"line_number,omitempty"`
	Text       string `json:"text"`
	Timestamp  int64  `json:"timestamp"`
	Branch     string `json:"branch"`
	Commit     string `json:"commit"`
	Resolved   bool   `json:"resolved"`
	ResolvedBy string `json:"resolved_by,omitempty"`
	ResolvedAt int64  `json:"resolved_at,omitempty"`
}

type Note struct {
	ID          string            `json:"id"`
	FilePath    string            `json:"file_path"`
	LineNumber  *int              `json:"line_number,omitempty"`
	Text        string            `json:"text"`
	Timestamp   int64             `json:"timestamp"`
	Branch      string            `json:"branch"`
	Commit      string            `json:"commit"`
	Author      string            `json:"author"` // e.g., "claude", "copilot", "human:username"
	Type        string            `json:"type"`   // e.g., "explanation", "rationale", "suggestion"
	Metadata    map[string]string `json:"metadata,omitempty"`
	Dismissed   bool              `json:"dismissed"`
	DismissedBy string            `json:"dismissed_by,omitempty"`
	DismissedAt int64             `json:"dismissed_at,omitempty"`
}

type RepoState struct {
	ViewedFiles []string   `json:"viewed_files"`
	Comments    []*Comment `json:"comments"`
	Notes       []*Note    `json:"notes"`
}

type ViewedState struct {
	Repos map[string]map[string]map[string]*RepoState `json:"repos"`
}

type Manager struct {
	stateFile      string
	state          *ViewedState
	exportBasePath string
}

func NewManager() (*Manager, error) {
	stateDir, err := getStateDir()
	if err != nil {
		return nil, err
	}

	if err := os.MkdirAll(stateDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create state directory: %w", err)
	}

	stateFile := filepath.Join(stateDir, "viewed.json")

	state := &ViewedState{
		Repos: make(map[string]map[string]map[string]*RepoState),
	}

	if _, err := os.Stat(stateFile); err == nil {
		data, err := os.ReadFile(stateFile)
		if err != nil {
			return nil, fmt.Errorf("failed to read state file: %w", err)
		}

		if err := json.Unmarshal(data, state); err != nil {
			// If unmarshal fails, start with empty state
			state = &ViewedState{
				Repos: make(map[string]map[string]map[string]*RepoState),
			}
		}
	}

	// Load config to get custom export path
	var exportBasePath string
	cfg, err := config.Load()
	if err == nil && cfg.ExportPath != "" {
		exportBasePath = cfg.ExportPath
	}

	return &Manager{
		stateFile:      stateFile,
		state:          state,
		exportBasePath: exportBasePath,
	}, nil
}

func (m *Manager) IsFileViewed(repoPath, branch, commit, filePath string) bool {
	if branches, ok := m.state.Repos[repoPath]; ok {
		if commits, ok := branches[branch]; ok {
			if repoState, ok := commits[commit]; ok {
				for _, viewed := range repoState.ViewedFiles {
					if viewed == filePath {
						return true
					}
				}
			}
		}
	}
	return false
}

func (m *Manager) MarkFileViewed(repoPath, branch, commit, filePath string) error {
	if m.state.Repos[repoPath] == nil {
		m.state.Repos[repoPath] = make(map[string]map[string]*RepoState)
	}

	if m.state.Repos[repoPath][branch] == nil {
		m.state.Repos[repoPath][branch] = make(map[string]*RepoState)
	}

	if m.state.Repos[repoPath][branch][commit] == nil {
		m.state.Repos[repoPath][branch][commit] = &RepoState{
			ViewedFiles: []string{},
			Comments:    []*Comment{},
			Notes:       []*Note{},
		}
	}

	repoState := m.state.Repos[repoPath][branch][commit]

	// Check if already viewed
	for _, viewed := range repoState.ViewedFiles {
		if viewed == filePath {
			return m.save()
		}
	}

	repoState.ViewedFiles = append(repoState.ViewedFiles, filePath)
	return m.save()
}

func (m *Manager) UnmarkFileViewed(repoPath, branch, commit, filePath string) error {
	if branches, ok := m.state.Repos[repoPath]; ok {
		if commits, ok := branches[branch]; ok {
			if repoState, ok := commits[commit]; ok {
				filtered := []string{}
				for _, viewed := range repoState.ViewedFiles {
					if viewed != filePath {
						filtered = append(filtered, viewed)
					}
				}
				repoState.ViewedFiles = filtered
			}
		}
	}

	return m.save()
}

func (m *Manager) AddComment(repoPath, branch, commit, filePath string, lineNumber *int, text string) (*Comment, error) {
	if m.state.Repos[repoPath] == nil {
		m.state.Repos[repoPath] = make(map[string]map[string]*RepoState)
	}

	if m.state.Repos[repoPath][branch] == nil {
		m.state.Repos[repoPath][branch] = make(map[string]*RepoState)
	}

	if m.state.Repos[repoPath][branch][commit] == nil {
		m.state.Repos[repoPath][branch][commit] = &RepoState{
			ViewedFiles: []string{},
			Comments:    []*Comment{},
			Notes:       []*Note{},
		}
	}

	repoState := m.state.Repos[repoPath][branch][commit]

	timestamp := time.Now().Unix()
	comment := &Comment{
		ID:         fmt.Sprintf("%d-%d", timestamp, len(repoState.Comments)),
		FilePath:   filePath,
		LineNumber: lineNumber,
		Text:       text,
		Timestamp:  timestamp,
		Branch:     branch,
		Commit:     commit,
		Resolved:   false,
	}

	repoState.Comments = append(repoState.Comments, comment)

	if err := m.save(); err != nil {
		return nil, err
	}

	return comment, nil
}

func (m *Manager) GetComments(repoPath, branch, commit string, filePath *string) []*Comment {
	if branches, ok := m.state.Repos[repoPath]; ok {
		if commits, ok := branches[branch]; ok {
			if repoState, ok := commits[commit]; ok {
				if filePath == nil {
					return repoState.Comments
				}

				filtered := []*Comment{}
				for _, comment := range repoState.Comments {
					if comment.FilePath == *filePath {
						filtered = append(filtered, comment)
					}
				}
				return filtered
			}
		}
	}

	return []*Comment{}
}

func (m *Manager) ResolveComment(repoPath, branch, commit, commentID, resolvedBy string) error {
	if branches, ok := m.state.Repos[repoPath]; ok {
		if commits, ok := branches[branch]; ok {
			if repoState, ok := commits[commit]; ok {
				for _, comment := range repoState.Comments {
					if comment.ID == commentID {
						comment.Resolved = true
						comment.ResolvedBy = resolvedBy
						comment.ResolvedAt = time.Now().Unix()
						return m.save()
					}
				}
			}
		}
	}

	return fmt.Errorf("comment not found")
}

func (m *Manager) GetAllComments(repoPath string) []*Comment {
	var allComments []*Comment

	if branches, ok := m.state.Repos[repoPath]; ok {
		for _, commits := range branches {
			for _, repoState := range commits {
				allComments = append(allComments, repoState.Comments...)
			}
		}
	}

	return allComments
}

func (m *Manager) AddNote(repoPath, branch, commit, filePath string, lineNumber *int, text, author, noteType string, metadata map[string]string) (*Note, error) {
	if m.state.Repos[repoPath] == nil {
		m.state.Repos[repoPath] = make(map[string]map[string]*RepoState)
	}

	if m.state.Repos[repoPath][branch] == nil {
		m.state.Repos[repoPath][branch] = make(map[string]*RepoState)
	}

	if m.state.Repos[repoPath][branch][commit] == nil {
		m.state.Repos[repoPath][branch][commit] = &RepoState{
			ViewedFiles: []string{},
			Comments:    []*Comment{},
			Notes:       []*Note{},
		}
	}

	repoState := m.state.Repos[repoPath][branch][commit]

	timestamp := time.Now().Unix()
	note := &Note{
		ID:         fmt.Sprintf("%d-%d", timestamp, len(repoState.Notes)),
		FilePath:   filePath,
		LineNumber: lineNumber,
		Text:       text,
		Timestamp:  timestamp,
		Branch:     branch,
		Commit:     commit,
		Author:     author,
		Type:       noteType,
		Metadata:   metadata,
		Dismissed:  false,
	}

	repoState.Notes = append(repoState.Notes, note)

	if err := m.save(); err != nil {
		return nil, err
	}

	return note, nil
}

func (m *Manager) GetNotes(repoPath, branch, commit string, filePath *string) []*Note {
	if branches, ok := m.state.Repos[repoPath]; ok {
		if commits, ok := branches[branch]; ok {
			if repoState, ok := commits[commit]; ok {
				if filePath == nil {
					return repoState.Notes
				}

				filtered := []*Note{}
				for _, note := range repoState.Notes {
					if note.FilePath == *filePath {
						filtered = append(filtered, note)
					}
				}
				return filtered
			}
		}
	}

	return []*Note{}
}

func (m *Manager) GetAllNotes(repoPath string) []*Note {
	var allNotes []*Note

	if branches, ok := m.state.Repos[repoPath]; ok {
		for _, commits := range branches {
			for _, repoState := range commits {
				allNotes = append(allNotes, repoState.Notes...)
			}
		}
	}

	return allNotes
}

func (m *Manager) DismissNote(repoPath, branch, commit, noteID, dismissedBy string) error {
	if branches, ok := m.state.Repos[repoPath]; ok {
		if commits, ok := branches[branch]; ok {
			if repoState, ok := commits[commit]; ok {
				for _, note := range repoState.Notes {
					if note.ID == noteID {
						note.Dismissed = true
						note.DismissedBy = dismissedBy
						note.DismissedAt = time.Now().Unix()
						return m.save()
					}
				}
			}
		}
	}

	return fmt.Errorf("note not found")
}

func (m *Manager) save() error {
	data, err := json.MarshalIndent(m.state, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to serialize state: %w", err)
	}

	if err := os.WriteFile(m.stateFile, data, 0644); err != nil {
		return fmt.Errorf("failed to write state file: %w", err)
	}

	if err := m.exportToJSON(); err != nil {
		log.Printf("Warning: failed to export JSON: %v", err)
	}

	return nil
}

func (m *Manager) exportToJSON() error {
	for repoPath, branches := range m.state.Repos {
		if err := m.exportRepoToJSON(repoPath, branches); err != nil {
			log.Printf("Warning: failed to export repo %s: %v", repoPath, err)
		}
	}
	return nil
}

func (m *Manager) exportRepoToJSON(repoPath string, branches map[string]map[string]*RepoState) error {
	exportPath, err := export.GetExportPathForRepoWithBase(repoPath, m.exportBasePath)
	if err != nil {
		return err
	}

	var comments []*export.Comment
	var notes []*export.Note

	for _, commits := range branches {
		for _, repoState := range commits {
			for _, c := range repoState.Comments {
				comments = append(comments, &export.Comment{
					ID:         c.ID,
					FilePath:   c.FilePath,
					LineNumber: c.LineNumber,
					Text:       c.Text,
					Timestamp:  c.Timestamp,
					Branch:     c.Branch,
					Commit:     c.Commit,
					Resolved:   c.Resolved,
					ResolvedBy: c.ResolvedBy,
					ResolvedAt: c.ResolvedAt,
				})
			}
			for _, n := range repoState.Notes {
				notes = append(notes, &export.Note{
					ID:          n.ID,
					FilePath:    n.FilePath,
					LineNumber:  n.LineNumber,
					Text:        n.Text,
					Timestamp:   n.Timestamp,
					Branch:      n.Branch,
					Commit:      n.Commit,
					Author:      n.Author,
					Type:        n.Type,
					Metadata:    n.Metadata,
					Dismissed:   n.Dismissed,
					DismissedBy: n.DismissedBy,
					DismissedAt: n.DismissedAt,
				})
			}
		}
	}

	return export.Export(repoPath, comments, notes, exportPath)
}

func getStateDir() (string, error) {
	// Use XDG_STATE_HOME on Unix, or fallback to XDG_DATA_HOME/LocalAppData
	if stateHome := os.Getenv("XDG_STATE_HOME"); stateHome != "" {
		return filepath.Join(stateHome, "guck"), nil
	}

	if dataHome := os.Getenv("XDG_DATA_HOME"); dataHome != "" {
		return filepath.Join(dataHome, "guck"), nil
	}

	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("failed to determine home directory: %w", err)
	}

	// Platform-specific defaults
	return filepath.Join(home, ".local", "state", "guck"), nil
}
